File: kits/core/spring-boot-toolkit-utils/src/main/java/org/toolkit/spring/boot/utils/bean/BeanUtil.java
package org.toolkit.spring.boot.utils.bean;

import java.lang.annotation.Annotation;
import java.util.*;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.beans.factory.support.DefaultListableBeanFactory;
import org.springframework.context.ApplicationContext;
import org.springframework.core.ResolvableType;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class BeanUtil {

	private final ApplicationContext context;
	private final DefaultListableBeanFactory factory;

	public BeanUtil(ApplicationContext context, DefaultListableBeanFactory factory) {
		this.context = context;
		this.factory = factory;
	}

	public final <T> List<T> getBeansOfType(Class<T> clazz) {
		return context.getBeansOfType(clazz).values().stream().toList();
	}

	public final <T> List<T> findBeans(Class<T> clazz) {
		return context.getBeansOfType(clazz).values().stream().toList();
	}

	@SafeVarargs
	public final <T> @Nullable T getBeansWithAnnotationAndGenericType(
			Class<T> clazz, Annotation ann, Class<T>... genericType) {
		ResolvableType.forClassWithGenerics(clazz, genericType);
		context.getBeansOfType(clazz);
		return null;
	}

	public final <T> void putAllAsSingleton(@NotNull Map<String, T> beans) {
		beans.forEach(factory::registerSingleton);
	}
}

File: kits/core/spring-boot-toolkit-utils/src/main/java/org/toolkit/spring/boot/utils/bean/BeanUtil.kt
package org.toolkit.spring.boot.utils.bean

import lombok.extern.slf4j.Slf4j
import org.jetbrains.annotations.Contract
import org.jetbrains.annotations.NotNull
import org.springframework.beans.factory.support.DefaultListableBeanFactory
import org.springframework.context.ApplicationContext
import org.springframework.core.ResolvableType
import org.springframework.stereotype.Component
import java.util.*

@Component
@Slf4j
class BeanUtil(private val context: ApplicationContext, private val beanFactory: DefaultListableBeanFactory) {
    fun <T> findBeans(clazz: Class<T>): List<T> {
        return context.getBeansOfType(clazz).values.stream().toList()
    }

    @Contract("_, _, _ -> new")
    fun <T> getBeansWithAnnotationAndGenericType(
        clazz: Class<*>?, annotation: Class<out Annotation?>?, vararg genericTypes: Class<*>?
    ): Map<String, T> {
        val type = clazz?.let { ResolvableType.forClassWithGenerics(it, *genericTypes) }
        annotation?.let { context.getBeansWithAnnotation(it) }
        return HashMap()
    }

    @SafeVarargs
    fun <T> getBeansWithGenericType(clazz: Class<T>?, vararg genericTypes: Class<T>?): Optional<T> {
        val type = clazz?.let { ResolvableType.forClassWithGenerics(it, *genericTypes) }
        return Optional.ofNullable<T & Any>(type?.let { context.getBeanProvider<T>(it).getIfAvailable() })
    }

    fun <T> getBeansOfType(tClass: Class<T>): List<T> {
        return context.getBeansOfType(tClass).values.stream().toList()
    }

    fun <T> putAllAsSingleton(@NotNull beans: Map<String?, T>) {
        beans.forEach { (key, value) ->
            key?.let { beanFactory.registerSingleton(it, value!!) }
        }
    }
}
